### PR TITLE
Feature/algorithmist girl/adding conditional flow to home page

### PR DESCRIPTION
**Pull request by lynguyxn**
### Administrative Info
[GitHub Issue #9]
Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[3-4 word description separated by dashes]`. Otherwise, please rename your branch and create a new PR.

### Changes
What changes did you make?
- Created Auth context
  - Used expo secure store to store and access the jwt token
  - Implemented API calls with YSC server to create `login` and `validate` functions in which `login` obtains the jwt token and `validate` ensures that the token is not expired
- Created User context
  - Stores user payload from login request for easier future access
- Created .env file to privately store server route

### Testing
How did you confirm your changes worked? 
- Used testing credentials provided by Akash to confirm login works as expected
- Confirmed User state receives user payload information

### Confirmation of Change 
Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

| [Logging in with test credentials] |
| --- |

https://user-images.githubusercontent.com/64985802/160950684-2233cfca-2111-46f1-8c00-7d3a888d90cf.mp4

the password is blocked from showing up on video lol

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (nicolaswl) [inline package-lock.json]: I don't think we should have a package-lock.json in the root of the project.

File: mobile/src/screens/HomeScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { Text, View } from "react-native";
+
+import Button from "../components/Button";
+import { AppStylesheet } from "../styles/AppStylesheet";
+
+const HomeScreen = () => (
+  <View style={AppStylesheet.container}>
+    <Text style={AppStylesheet.headerHomeScreen}>Welcome, User</Text>
```
Comment (nicolaswl) [inline mobile/src/screens/HomeScreen.tsx]: How hard would it be to actually specify the user here? Can we grab it from the context easily?

File: mobile/src/contexts/AuthContext.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        const res = await fetch(url, {
+          method: "POST",
+        });
+
+        if (res.ok) {
+          const tokenJson = await res.json();
+          const token = JSON.stringify(tokenJson);
+          // store token & user info
+          await SecureStore.setItemAsync("token", token);
+          const decoded: any = jwt_decode(token);
```
Comment (nicolaswl) [inline mobile/src/contexts/AuthContext.tsx]: Why are we specifying the type as any here?

File: mobile/src/contexts/AuthContext.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            username: decoded.username,
+            firstName: decoded.firstName,
+            lastName: decoded.lastName,
+            role: decoded.role,
+            email: decoded.email,
+          };
+          setUserState(newUserState);
+          setIsLoggedIn(true);
+        } else {
+          console.error("Login request was unsuccessful.");
```
Comment (nicolaswl) [inline mobile/src/contexts/AuthContext.tsx:63]: If we want to keep these for logging purpose, make sure to disable the lint rule with a comment. Here are some examples https://chaseadams.io/posts/eslint-disable-rule-by-comment-generator/

File: mobile/src/contexts/AuthContext.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+type AuthState = {
+  isLoggedIn: boolean;
+  login: (username: string, password: string) => void;
+  validate: () => void;
+};
+
+const initialState: AuthState = {
+  isLoggedIn: false,
+  login: () => {
+    // initial function
```
Comment (nicolaswl) [inline mobile/src/contexts/AuthContext.tsx]: Why do we need to do this in the initial states?

**Review by nicolaswl — COMMENTED** [review]
This looks really solid! I just left some questions and asked for some pretty simple changes. Did we figure out why the package-lock diffs are so huge? Maybe try deleting them and remaking them with `npm i` and see if that brings them down?

File: mobile/src/screens/HomeScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+    <Button text="Play Game With A Mentor" />
```
Comment (nicolaswl) [inline mobile/src/screens/HomeScreen.tsx]: We only need `Play Game` here for MVP. Here's the MVP figma https://www.figma.com/file/gRuvdstnqbXk5RbzgTQsTN/YSC-Mobile-Application?node-id=114%3A42

File: mobile/src/screens/HomeScreen.tsx (hunk 2 above)
Comment (lynguyxn) [reply mobile/src/screens/HomeScreen.tsx]: yeah, we can easily grab it using `const { userState } = useContext(UserContext);` 
do you want it to display the user's first name?

File: package-lock.json (hunk 1 above)
Comment (lynguyxn) [reply package-lock.json]: resolved

File: mobile/src/contexts/AuthContext.tsx (hunk 3 above)
Comment (lynguyxn) [reply mobile/src/contexts/AuthContext.tsx]: resolved, created a new interface `Payload`

File: mobile/src/contexts/AuthContext.tsx (hunk 4 above)
Comment (lynguyxn) [reply mobile/src/contexts/AuthContext.tsx:63]: resolved

File: mobile/src/screens/HomeScreen.tsx (hunk 2 above)
Comment (nicolaswl) [reply mobile/src/screens/HomeScreen.tsx]: Yeah I think first name is fine for now!

**Review by nicolaswl — APPROVED** [review]
LGTM

**Review by rohanputhukudy — APPROVED** [review]
Looks good to me!

**Pull request merged**